src/Order/Checkout/CheckoutForm.php

```diff
@@ -0,0 +1,269 @@
+        $defaultMessage = _t(self::class . '.PAYMENT_ERROR',
+            'There was an error processing your payment. Please try again.');
+
+        if ($payment === null) {
+            // No payment with that identifier for this order, can't show error
+            $errorMessage = false;
+        } elseif ($payment->Status === PaymentStatus::VOID) {
+            // Void status is returned for cancelled or declined card
+            $errorMessage = $defaultMessage;
+        } else {
+            /** @var PaymentMessage|null $message */
+            $message = $payment->Messages()->last();
+            $errorMessage = $message instanceof PurchaseError || $message instanceof CompletePurchaseError
+                ? $message->Message
+                : $defaultMessage;
+        }
+
+        $this->paymentError = $errorMessage;
+        return $errorMessage ?: null;
+    }
+
+    /**
+     * @param array $data
+     * @return HTTPResponse
+     * @throws \SilverStripe\Omnipay\Exception\InvalidConfigurationException
+     * @throws \SilverStripe\Omnipay\Exception\InvalidStateException
+     * @throws \Exception
+     */
+    public function ConfirmCheckout(array $data): HTTPResponse
+    {
+        $this->cart->Lock();
+        $this->extend('beforeInitPayment', $data);
+
+        $this->saveInto($this->cart);
+        $this->cart->write();
+
+        /** @var Payment|PaymentExtension $payment */
+        $payment = Payment::create();
+        $payment->OrderID = $this->cart->ID;
+
+        $paymentMethod = $data[static::PAYMENT_METHOD_FIELD];
+        $dueMoney = $this->cart->UnpaidTotal()->getMoney();
+        $payment->init($paymentMethod, $this->supportedCurrencies->formatDecimal($dueMoney),
+            $dueMoney->getCurrency()->getCode())
+            ->setSuccessUrl($this->getSuccessUrl())
+            ->setFailureUrl($this->getFailureUrl($payment));
+
+        if ($payment->isChanged(null, Payment::CHANGE_VALUE)) {
+            $payment->write();
+        }
+
+        $response = $this->paymentServiceFactory
+            ->getService($payment, ServiceFactory::INTENT_PURCHASE)
+            ->initiate($data);
+
+        $this->extend('afterInitPayment', $data, $payment, $response);
+        return $response->redirectOrRespond();
+    }
+
+    /**
+     * @return string
+     */
+    protected function getSuccessUrl(): string
+    {
+        /** @var OrderConfirmationPage $orderConfirmationPage */
+        $orderConfirmationPage = OrderConfirmationPage::get_one(OrderConfirmationPage::class);
+        return $orderConfirmationPage->LinkForOrder($this->cart);
+    }
+
+    /**
+     * @param Payment $payment
+     * @return string
+     * @throws \SilverStripe\ORM\ValidationException
+     */
+    protected function getFailureUrl(Payment $payment): string
+    {
+        if (!$payment->Identifier) {
+            // Force identifier to be generated
+            /** @see Payment::onBeforeWrite() */
+            $payment->write();
```

Comment:
Shouldn't you do this in Payment::init() so you don't need to check it later?

Reply:
This is an issue/limitation of the `silverstripe/silverstripe-omnipay` module.